<?xml version="1.0"?>
<launch>
  <master auto="start"/>
  <arg name="model" default="$(find p3dx_description)/urdf/pioneer3dx.xml"/>
  <param name="robot_description" textfile="$(arg model)" />
  <param name="/use_sim_time" value="true"/>

  <node pkg="stage_ros" type="stageros" name="stageros"
args="$(find co-coverage)/launch/demo_5.world" respawn="false">
    <param name="base_watchdog_timeout" value="0.2"/>
  </node>

<!--- Run map server and rviz -->
<node name="map_server" pkg="map_server" type="map_server"
  args="$(find co-coverage)/launch/wild.pgm 0.05" respawn="false"/>
<node pkg="rviz" type="rviz" name="rviz" args="-d $(find co-coverage)/rviz_configs/stage_multi_p3dx-wild.rviz"/>

 <!-- LEADER ROBOT -->
<group ns="robot_0">
 <arg name="tf_prefix" value="robot_0" />
 <node pkg="co-coverage" type="rf_node_co_coverage" name="rf_node_co_coverage">
    <param name="instampname" value="/robot_0" />
    <param name="idstamp" value="0" />
    <param name="inleaderstampname" value="/robot_0" />
    <param name="inlid" value="0.70711" />
    <param name="inlambda" value="1.0" />
    <param name="intau_delta" value="3.0" />
    <param name="inmu" value="1.0" />
    <param name="ingamma" value="PI/4" />
    <param name="inangle_desired" value="-0.40" />
    <param name="incvrep" value="0.9" />
    <param name="incvatt" value="0.5" />
    <remap from="scan" to="base_scan"/>
  </node>
 
 <node pkg="teleop_twist_keyboard" type="teleop_twist_keyboard.py" name="teleop" output="screen"> 
 </node>

 <include file="$(find p3dx_description)/launch/p3dx_description.launch">
   <arg name="tf_prefix" value="$(arg tf_prefix)" />
  </include>

  <node pkg="fake_localization" type="fake_localization" name="fake_localization" output="screen">
    <param name="global_frame_id" value="map"/>
    <param name="odom_frame_id" value="$(arg tf_prefix)/odom"/>
    <param name="base_frame_id" value="$(arg tf_prefix)/base_footprint"/>
  </node>
</group>

<!-- FOLLOWER ROBOTS -->
<group ns="robot_1">
  <arg name="tf_prefix" value="robot_1" />
  <node pkg="co-coverage" type="rf_node_co_coverage" name="rf_node_co_coverage">
    <param name="instampname" value="/robot_1" />
    <param name="idstamp" value="1" />
    <param name="inleaderstampname" value="/robot_0" />
    <param name="inlid" value="0.70711" />
    <param name="inlambda" value="1.0" />
    <param name="intau_delta" value="3.0" />
    <param name="inmu" value="1.0" />
    <param name="ingamma" value="PI/4" />
    <param name="inangle_desired" value="-0.40" />
    <param name="incvrep" value="0.9" />
    <param name="incvatt" value="0.5" />
    <remap from="scan" to="base_scan"/>
  </node>

  <include file="$(find p3dx_description)/launch/p3dx_description.launch">
   <arg name="tf_prefix" value="$(arg tf_prefix)" />
  </include>

  <node pkg="fake_localization" type="fake_localization" name="fake_localization" output="screen">
    <param name="global_frame_id" value="map"/>
    <param name="odom_frame_id" value="$(arg tf_prefix)/odom"/>
    <param name="base_frame_id" value="$(arg tf_prefix)/base_footprint"/>
  </node>

</group>

<group ns="robot_2">
  <arg name="tf_prefix" value="robot_2" />
  <node pkg="co-coverage" type="rf_node_co_coverage" name="rf_node_co_coverage">
    <param name="instampname" value="/robot_2" />
    <param name="idstamp" value="2" />
    <param name="inleaderstampname" value="/robot_1" />
    <param name="inlid" value="0.70711" />
    <param name="inlambda" value="1.0" />
    <param name="intau_delta" value="3.0" />
    <param name="inmu" value="1.0" />
    <param name="ingamma" value="PI/4" />
    <param name="inangle_desired" value="-0.40" />
    <param name="incvrep" value="0.9" />
    <param name="incvatt" value="0.5" />
    <remap from="scan" to="base_scan"/>
  </node>

  <include file="$(find p3dx_description)/launch/p3dx_description.launch">
   <arg name="tf_prefix" value="$(arg tf_prefix)" />
  </include>

  <node pkg="fake_localization" type="fake_localization" name="fake_localization" output="screen">
    <param name="global_frame_id" value="map"/>
    <param name="odom_frame_id" value="$(arg tf_prefix)/odom"/>
    <param name="base_frame_id" value="$(arg tf_prefix)/base_footprint"/>
  </node>

</group>

<group ns="robot_3">
  <arg name="tf_prefix" value="robot_3" />
  <node pkg="co-coverage" type="rf_node_co_coverage" name="rf_node_co_coverage">
    <param name="instampname" value="/robot_3" />
    <param name="idstamp" value="3" />
    <param name="inleaderstampname" value="/robot_2" />
    <param name="inlid" value="0.70711" />
    <param name="inlambda" value="1.0" />
    <param name="intau_delta" value="3.0" />
    <param name="inmu" value="1.0" />
    <param name="ingamma" value="PI/4" />
    <param name="inangle_desired" value="-0.40" />
    <param name="incvrep" value="0.9" />
    <param name="incvatt" value="0.5" />
    <remap from="scan" to="base_scan"/>
  </node>

  <include file="$(find p3dx_description)/launch/p3dx_description.launch">
   <arg name="tf_prefix" value="$(arg tf_prefix)" />
  </include>

  <node pkg="fake_localization" type="fake_localization" name="fake_localization" output="screen">
    <param name="global_frame_id" value="map"/>
    <param name="odom_frame_id" value="$(arg tf_prefix)/odom"/>
    <param name="base_frame_id" value="$(arg tf_prefix)/base_footprint"/>
  </node>

</group>

<group ns="robot_4">
  <arg name="tf_prefix" value="robot_4" />
  <node pkg="co-coverage" type="rf_node_co_coverage" name="rf_node_co_coverage">
    <param name="instampname" value="/robot_4" />
    <param name="idstamp" value="4" />
    <param name="inleaderstampname" value="/robot_3" />
    <param name="inlid" value="0.70711" />
    <param name="inlambda" value="1.0" />
    <param name="intau_delta" value="3.0" />
    <param name="inmu" value="1.0" />
    <param name="ingamma" value="PI/4" />
    <param name="inangle_desired" value="-0.40" />
    <param name="incvrep" value="0.9" />
    <param name="incvatt" value="0.5" />
    <remap from="scan" to="base_scan"/>
  </node>

  <include file="$(find p3dx_description)/launch/p3dx_description.launch">
   <arg name="tf_prefix" value="$(arg tf_prefix)" />
  </include>

  <node pkg="fake_localization" type="fake_localization" name="fake_localization" output="screen">
    <param name="global_frame_id" value="map"/>
    <param name="odom_frame_id" value="$(arg tf_prefix)/odom"/>
    <param name="base_frame_id" value="$(arg tf_prefix)/base_footprint"/>
  </node>
  
</group>

</launch>
